virtual_rainforest/models/abiotic/radiation.py

```diff
@@ -0,0 +1,317 @@
+    topofcanopy_radiation: NDArray[np.float32] = field(init=False)
+    """Top of canopy downward shortwave radiation, [J m-2]"""
+    longwave_radiation: NDArray[np.float32] = field(init=False)
+    """Longwave radiation from canopy layers and soil, [J m-2]"""
+    netradiation_surface: NDArray[np.float32] = field(init=False)
+    """Net shortwave radiation at the surface (= forest floor), [J m-2]"""
+
+    def __post_init__(self) -> None:
+        """Post init method populates radiation balance attributes from data object.
+
+        Args:
+            elevation: elevation above sea level, [m]
+            shortwave_in: downward shortwave radiation, [J m-2]
+            canopy_temperature: canopy temperature of individual layers, [C]
+            surface_temperature: surface soil temperature, [C]
+            sunshine_fraction: fraction of sunshine hours, between 0 (100% cloud cover)
+                and 1 (cloud free sky), default = 1
+            canopy_absorption: absorption by canopy, [J m-2]
+            albedo_vis: visible light albedo, default = 0.03
+            albedo_shortwave: shortwave albedo, default = 0.17
+        """
+
+        self.ppfd = calculate_ppfd(data["shortwave_in"], data["elevation"])
+        self.topofcanopy_radiation = calculate_topofcanopy_radiation(
+            data["shortwave_in"], data["elevation"]
+        )
+        self.longwave_radiation = calculate_longwave_radiation(
+            data["canopy_temperature"], data["surface_temperature"]
+        )
+        self.netradiation_surface = calculate_netradiation_surface(
+            self.topofcanopy_radiation,
+            data["canopy_absorption"],
+            self.longwave_radiation,
+        )
+
+
+# helper functions
+def calculate_atmospheric_transmissivity(
+    elevation: NDArray[np.float32],
+    sunshine_fraction: NDArray[np.float32] = np.array(1.0, dtype=np.float32),
+    cloudy_transmissivity: float = CLOUDY_TRANSMISSIVITY,
+    transmissivity_coefficient: float = TRANSMISSIVITY_COEFFICIENT,
+    beer_regression: float = BEER_REGRESSION,
+) -> NDArray[np.float32]:
+    """Calculate atmospheric transmissivity (tau).
+
+    Args:
+        elevation: elevation above sea level, [m]
+        sunshine_fraction: fraction of sunshine hours, between 0 (100% cloud cover)
+            and 1 (cloud free sky), default = 1
+        cloudy_transmissivity: cloudy transmittivity :cite:p:`Linacre1968`, default set
+            in config
+        transmissivity_coefficient: angular coefficient of transmittivity
+            :cite:p:`Linacre1968`, default set in config
+        beer_regression: parameter in equation for atmospheric transmissivity based on
+            regression of Beer's radiation extinction function :cite:p:`Allen1996`,
+            default set in config
+
+    Returns:
+        atmospheric transmissivity, unitless
+    """
+
+    # check sunshine fraction between 0 and 1
+    if 0 > np.any(sunshine_fraction) > 1:
```

Comment:
There are two problems here:

1. Doesn't this need to be between 0 and 1 *inclusive*? Currently you can't have 0% or 100% sunshine
2. The syntax is wrong. This will check whether any of `sunshine_fraction` is "truthy" first, *then* it will check whether this true/false value is between 0 and 1

I think this will work:

```suggestion
    if np.any(0 => sunshine_fraction => 1):
```

Reply:
We might want to turn this into a helper function because I think it will get widely used. 😄 Once we get the data input bounds checking setup (#135) that could include a basic function we could import to do this.